[processor/metricstransform] Migrate the processor from OC to pdata

dmitryax opened the pull request:
As side effect this change:
- Reduces CPU/memory utilization
- Fixes issues caused by OC<->OTLP translation, e.g. undesired conversion of non-monothonic sum metrics to gauges.
- Scope info (name + version) is not dropped anymore
- Adds support of exponential histograms
- Adds ability to aggregate numeric data points with different value type (int, double)

This change can be reverted by disabling the `processor.metricstransformprocessor.UseOTLPDataModel` feature gate (add `--feature-gates=-processor.metricstransformprocessor.UseOTLPDataModel` command line argument).

Benchmark results:
```
goos: darwin
goarch: amd64
pkg: github.com/open-telemetry/opentelemetry-collector-contrib/processor/metricstransformprocessor
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
```

before:
```
BenchmarkMetricsTransformProcessorRenameMetrics-16    	    2390	    487938 ns/op
```

after:
```
BenchmarkMetricsTransformProcessorRenameMetrics-16    	   11220	    105151 ns/op
```

Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10575

Also fixes 
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/10420
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/7071
- https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/4390

dmitryax commented:
@open-telemetry/collector-approvers does anyone have a chance to review this PR?

dmitryax commented:
@open-telemetry/collector-contrib-approvers friendly ping. PTAL

dmitryax commented:
cc @punya as code owner

TylerHelmuth reviewed (commented):
Read through everything today, but still pretty hard to tell if the new changes work the same way as the old one.  Any chances this can be broken into smaller PRs, even if the changes aren't used until some final PR that switches the processor over to OTLP?

dmitryax commented on processor/metricstransformprocessor/metrics_transform_processor_test.go:
```diff
@@ -75,78 +78,10 @@ func TestMetricsTransformProcessor(t *testing.T) {
 	}
 }
 
-func TestComputeDistVals(t *testing.T) {
```
This was testing a translation of an OC histogram's field `SumOfSquaredDeviation`. But that field isn't supported by OTLP, it's just ignored during [OC<->OTLP translation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/15b930b99dd777eba20391fbf5b5ece3cda0ff8d/pkg/translator/opencensus/metrics_to_oc.go#L271). So after the collector was switched to OTLP data model pipeline this became a dead code. This is probably the only part that can easily be extracted into a separate PR.

dmitryax commented on processor/metricstransformprocessor/metrics_transform_processor.go:
```diff
@@ -212,175 +310,268 @@ func newMetricsTransformProcessor(logger *zap.Logger, internalTransforms []inter
-					// TODO: report via trace / metric instead
-					mtp.logger.Warn(err.Error())
+				matchedMetrics := matchMetrics(transform.MetricIncludeFilter, metrics)
+				if len(matchedMetrics) == 0 {
```
Good point!

dmitryax commented on processor/metricstransformprocessor/operation_toggle_scalar_datatype.go:
```diff
@@ -14,28 +14,32 @@
-				dp.Value = &metricspb.Point_Int64Value{Int64Value: int64(dp.GetDoubleValue())}
-			}
-		}
+func toggleScalarDataTypeOp(metric pmetric.Metric, f internalFilter) {
```
done

dmitryax commented on processor/metricstransformprocessor/operation_add_label.go:
```diff
@@ -14,18 +14,14 @@
-		}
-		ts.LabelValues = append(ts.LabelValues, lv)
-	}
+func addLabelOp(metric pmetric.Metric, op internalOperation) {
```
done

bogdandrutu commented:
@dmitryax @TylerHelmuth I would encourage to have this protected by a featuregate (enabled by default) just in case there is a bug and users can switch back to the previous code. We should keep that for few versions.

dmitryax commented:
Sounds good! I'll change it use to a feature gate

dmitryax commented:
@TylerHelmuth I also move some code to separate PRs: 
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/11045
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/11047

dmitryax committed:
[processor/metricstransform] Migrate the processor from OC to pdata

As side effect this change:

- Reduces CPU/memory utilization
- Fixes issues caused by OC<->OTLP translation, e.g. undesired conversion of non-monothonic sum metrics to gauges.
- Adds support of exponential histograms
- Adds ability to aggregate numeric data points with different value type (int, double)

This change can be reverted by setting the `processor.metricstransformprocessor.UseOTLPDataModel` feature gate to "false".

dmitryax commented:
@TylerHelmuth @bogdandrutu I changed the PR to use a feature gate. PTAL

TylerHelmuth commented on processor/metricstransformprocessor/metrics_transform_processor.go:
```diff
@@ -25,17 +25,38 @@ import (
+	Enabled: true,
+	Description: "Controls which data model the processor uses for metric transformation internally: the new OLTP " +
+		"data model if set to \"true\", or the old OpenCensus data model if set to \"false\". " +
+		"If you see any bugs in the processor introduced after 0.34.0 version, " +
```
Should this be 0.54.0?

dmitryax replied:
Right, good catch :) fixed

dmitryax commented:
@open-telemetry/collector-maintainers I believe this PR is ready to merge. Good that it's enabled through a feature gate, so it can be reverted in case of any problems
